source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm

```diff
@@ -7,112 +7,212 @@
+/// maintain the sorted order after the removal
+- (void)removeVisibleViewAtIndex:(NSUInteger)index
+{
+    NSNumber *indexAsNumber = [NSNumber numberWithLong:index];
+    if (![_visibleViews containsObject:indexAsNumber]) {
+        return;
     }
+    NSRange range = NSMakeRange(0, _visibleViews.count);
+    NSUInteger removalIndex = [_visibleViews indexOfObject:indexAsNumber
+                                             inSortedRange:range
+                                                   options:NSBinarySearchingInsertionIndex
+                                           usingComparator:^(id num0, id num1) {
+                                               return [num0 compare:num1];
+                                           }];
+    [_visibleViews removeObjectAtIndex:removalIndex];
+}
 
-    // get the index of the target view
-    // if the target view is not the tail,
-    // check to the right, and if the right separator and not hidden,
-    // hide the separator
-    // if the target view is tail, check to left, and if the left is separator and not hidden,
-    // hide the separator
-    NSInteger index = [subviews indexOfObject:viewToBeHidden];
-    if (index + 1 < subviews.count) {
-        NSInteger nextViewIndex = index + 1;
-        UIView *nextView = subviews[nextViewIndex];
-        if ([nextView isKindOfClass:[ACRSeparator class]]) {
-            if (!nextView.isHidden) {
-                nextView.hidden = YES;
-            }
-        }
-    } else if (index - 1 >= 0) {
-        NSInteger prevViewIndex = index - 1;
-        UIView *prevView = subviews[prevViewIndex];
-        if ([prevView isKindOfClass:[ACRSeparator class]]) {
-            if (!prevView.isHidden) {
-                prevView.hidden = YES;
-            }
-        }
+/// YES means the index of view is currently the leading or top view
+- (BOOL)amIHead:(NSUInteger)index
+{
+    NSNumber *indexAsNumber = [NSNumber numberWithLong:index];
+    return (_visibleViews.count && [_visibleViews.firstObject isEqual:indexAsNumber]);
+}
+
+/// returns the current leading or top view's index
+- (NSUInteger)getHeadIndexOfVisibleViews
+{
+    if (_visibleViews.count) {
+        return [_visibleViews.firstObject longValue];
     }
 
-    // only cares for tailing padding at the moment
-    // if column width is "auto", and all of its children hidden,
-    // hide column, if is "stretch", then don't hide it.
-    if (self.padding && [_columnWidth isEqualToString:@"auto"]) {
-        for (UIView *subview in subviews) {
-            if (!subview.isHidden && ![subview isEqual:self.padding]) {
-                return;
-            }
+    return NSNotFound;
+}
+
+/// change the visibility of the separator of a host view to `visibility`
+/// `visibility` `YES` indicates that the separator will be hidden
+- (void)changeVisiblityOfSeparator:(UIView *)hostView visibilityValue:(BOOL)visibility contentStackView:(ACRContentStackView *)contentStackView
```

Comment:
>visibility [](http://example.com/codeflow?start=76&length=10)

I feel like visibility is tied to the possibility of seeing/detecting an element, it's odd to have it tied to the hidden value, this could be fixed by negating the visibility value or renaming the parameter

Reply:
The name of the visibility property of UIView is hidden.
If Set YES, the view will be hidden.

Reply:
fair enough